lib/Core/Executor.cpp

```diff
@@ -4006,32 +3986,38 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
       PersistentState *ps = dyn_cast<PersistentState>(wos);
       ref<Expr> offset = mo->getOffsetExpr(address);
 
-      ref<Expr> alreadyPersisted = ps->isOffsetAlreadyPersisted(offset);
-      StatePair notPersisted = fork(state, Expr::createIsZero(alreadyPersisted) , true);
-      if (notPersisted.first) {
-        auto& goodState = *notPersisted.first;
-        os = goodState.addressSpace.findObject(mo);
-        assert(os && "Could not get ObjectState from notPersisted.first");
-        wos = goodState.addressSpace.getWriteable(mo, os);
-        assert(wos && "Could not get writeable ObjectState from notPersisted.first");
-        ps = dyn_cast<PersistentState>(wos);
-        assert(ps && "Could not get PersistentState from notPersisted.first");
-        ps->persistCacheLineAtOffset(offset);
-        llvm::errs() << "Good Flush\n";
-      }
-      if (notPersisted.second) {
-        // offset is already persisted, should error!
-        auto& errState = *notPersisted.second;
-        os = errState.addressSpace.findObject(mo);
-        assert(os && "Could not get ObjectState from notPersisted.second");
-        wos = errState.addressSpace.getWriteable(mo, os);
-        assert(wos && "Could not get writeable ObjectState from notPersisted.second");
+      // Check if offset already flushed.
+      ref<Expr> check = ps->getIsOffsetPersistedExpr(offset,
+                                                     true /* allow pending */);
+
+      StatePair isAlreadyPersisted = fork(state, check, true);
+
+      auto *errState = isAlreadyPersisted.first;
+      auto *goodState = isAlreadyPersisted.second;
+
+      // Warn if already persisted.
+      if (errState) {
+        os = errState->addressSpace.findObject(mo);
+        assert(os);
+        wos = errState->addressSpace.getWriteable(mo, os);
+        assert(wos);
         ps = dyn_cast<PersistentState>(wos);
-        assert(ps && "Could not get PersistentState from notPersisted.second");
+        assert(ps);
         llvm::errs() << "Unnecessary Flush\n";
-        std::unordered_set<std::string> errors {ps->getLocationInfo(errState)};
         // avoid masking later bugs; emit error, but continue
-        emitPmemError(errState, errors);
+        emitPmemError(*errState, { ps->getLocationInfo(*errState) });
```

Comment:
should we also still invoke ps->persistCacheLineAtOffset() in the errState, since the state doesn't terminate?

Reply:
Well, it's already persisted :P